pkg/config/crd/crdwatcher.go

```diff
@@ -0,0 +1,273 @@
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp" // for k8s config for GKE
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"istio.io/mixer/pkg/config/store"
+)
+
+type resource struct {
+	version string
+	spec    interface{}
+}
+
+// Mapping from the resource name to the resource spec.
+type resourceMap map[string]*resource
+
+type watcher struct {
+	namespace string
+	stoppers  []func()
+
+	mu       sync.Mutex
+	data     map[string]string
+	listener store.Listener
+
+	// kind to resourceMap
+	resources map[string]resourceMap
+}
+
+var _ store.ChangeNotifier = &watcher{}
+
+// newStore creates a new watcher instance with watching resources.
+func newStore(u *url.URL) (store.KeyValueStore, error) {
+	namespace := u.Host
+	if namespace == "" {
+		namespace = v1.NamespaceDefault
+	}
+	configFilePath := u.Path
+	if configFilePath == "" {
+		configFilePath = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+	}
+	config, err := clientcmd.BuildConfigFromFlags("", configFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &watcher{
+		data:      map[string]string{},
+		namespace: namespace,
+		resources: map[string]resourceMap{},
+	}
+	if err = w.init(config); err != nil {
+		return nil, err
+	}
+	return w, nil
+}
+
+func (w *watcher) String() string {
+	return fmt.Sprintf("k8s: %s", w.namespace)
+}
+
+// Get implements a KeyValueStore method.
+func (w *watcher) Get(key string) (string, int, bool) {
+	w.mu.Lock()
+	value, found := w.data[key]
+	w.mu.Unlock()
+	return value, store.IndexNotSupported, found
+}
+
+// Set implements a KeyValueStore method.
+func (w *watcher) Set(key, value string) (int, error) {
+	return store.IndexNotSupported, errors.New("not implemented")
+}
+
+// List implements a KeyValueStore method.
+func (w *watcher) List(key string, recurse bool) ([]string, int, error) {
+	w.mu.Lock()
+	if !strings.HasSuffix(key, "/") {
+		key = key + "/"
+	}
+	keys := []string{}
+	for k := range w.data {
+		if strings.HasPrefix(k, key) {
+			keys = append(keys, k)
+		}
+	}
+	w.mu.Unlock()
+	return keys, store.IndexNotSupported, nil
+}
+
+// Delete implements a KeyValueStore method.
+func (w *watcher) Delete(key string) error {
+	return errors.New("not implemented")
+}
+
+func unstructuredToResource(uns *unstructured.Unstructured) (string, *resource) {
+	md := uns.Object["metadata"].(map[string]interface{})
+	name := md["name"].(string)
+	version := md["resourceVersion"].(string)
+	return name, &resource{version, uns.Object["spec"]}
+}
+
+func (w *watcher) appendAndEncode(data resourceMap, field, key string) {
```

Comment:
Please rename `data` to something else - maybe `resources`. It's confusing in context compared to `w.data`, especially because access to `w.data` needs to be guarded.